### PR TITLE
Initial import

### PR DESCRIPTION
**Pull request by basiliscos**

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -0,0 +1,14 @@
+sudo: false
+language: perl
+perl:
+    - "5.22"
+    - "5.20"
+    - "5.18"
+    - "5.16"
+    - "5.14"
+    - "5.12"
+    - "5.10"
```
Comment (torsten-deriv) [inline .travis.yml:10]: Do we really want this list?

File: Makefile.PL [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+use warnings;
+use ExtUtils::MakeMaker;
+
+WriteMakefile(
+    NAME             => 'CSV::HistoryPlayer',
+    AUTHOR           => q{binary.com <perl@binary.com>},
+    VERSION_FROM     => 'lib/CSV/HistoryPlayer.pm',
+    ABSTRACT         => 'Plays scattered CSV files with historic data',
+    LICENSE          => 'artistic_2',
+    MIN_PERL_VERSION => '5.8.0',
```
Comment (torsten-deriv) [inline Makefile.PL]: You don't really mean that, do you?

File: Makefile.PL [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    clean            => {FILES => 'CSV-HistoryPlayer-*'},
+    (
+        eval { ExtUtils::MakeMaker->VERSION(6.64) }
```
Comment (torsten-deriv) [inline Makefile.PL]: I'd rather bail out if this is not met. We don't care about ancient perls.

File: lib/CSV/HistoryPlayer.pm [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+    required => 1
+);
+
+has 'dir_filter' => (
+    is      => 'ro',
+    default => sub {
+        return sub { 1 }
+    });
+
+has 'files_maper' => (
```
Comment (torsten-deriv) [inline lib/CSV/HistoryPlayer.pm]: is this a typo?

File: lib/CSV/HistoryPlayer.pm [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+    my @files;
+    my @dirs_queue = (path($self->root_dir));
+    my $dir_filter = $self->dir_filter;
+    while (@dirs_queue) {
+        my $dir = shift(@dirs_queue);
+        if ($dir_filter->($dir)) {
+            for my $c ($dir->children) {
+                push @dirs_queue, $c if (-d $c);
+                push @files, $c
+                    if ($c =~ /\.csv$/i && -s $c && -r $c && -f $c);
```
Comment (torsten-deriv) [inline lib/CSV/HistoryPlayer.pm]: no need to `stat` the entry all over again. You did that before in line 49. Here you can simply write

``` perl
-s _ && -r _ && -f _
```

or even

``` perl
-s -r -f _
```

File: lib/CSV/HistoryPlayer.pm [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+        if (@ordered_idx) {
+            my $idx  = shift @ordered_idx;
+            my $line = $lines[$idx];
+            my $file = $cluser_files[$idx];
+            $self->_current_data([$file, $line]);
+            $lines[$idx] = undef;
+        } else {
+            if ($cluster_idx < @$clusters - 1) {
+                $open_cluster->(++$cluster_idx);
+                goto REDO;
```
Comment (torsten-deriv) [inline lib/CSV/HistoryPlayer.pm]: Just a style issue, could you please convert this to something like:

``` perl
REDO: {
    if () {
    } else {
        ...;
        redo REDO;
    }
}
```

or

``` perl
my $read_line_from_cluster;
$read_line_from_cluster = sub {
    ....
    goto &$read_line_from_cluster;
};
```

Yet, I'd prefer the former. I think the latter can create a memory leak.

File: lib/CSV/HistoryPlayer.pm.tidyup [hunk 7]
```diff
@@ -0,0 +1,132 @@
+package CSV::HistoryPlayer;
```
Comment (torsten-deriv) [inline lib/CSV/HistoryPlayer.pm.tidyup]: please remove this file

File: .travis.yml (hunk 1 above)
Comment (basiliscos) [inline .travis.yml:10]: Why not? This is open-source module, let it be tested on different perl versions

File: Makefile.PL (hunk 2 above)
Comment (basiliscos) [inline Makefile.PL]: Changed to `5.10` :)

File: lib/CSV/HistoryPlayer.pm (hunk 4 above)
Comment (basiliscos) [inline lib/CSV/HistoryPlayer.pm]: fixed

File: lib/CSV/HistoryPlayer.pm (hunk 5 above)
Comment (basiliscos) [inline lib/CSV/HistoryPlayer.pm]: I suspected that something like that exists, but I didn't know about `_` magic :)

File: lib/CSV/HistoryPlayer.pm (hunk 6 above)
Comment (basiliscos) [inline lib/CSV/HistoryPlayer.pm]: done

File: lib/CSV/HistoryPlayer.pm.tidyup (hunk 7 above)
Comment (basiliscos) [inline lib/CSV/HistoryPlayer.pm.tidyup]: done

**Pull request merged**